### ISSUE TITLE
Consensus (bank hash) mismatch between v1.8.6 and v1.8.7 (testnet)

### ISSUE THREAD

**Issue by steviez** [solana-labs/solana#21850]
#### Problem
A transaction was found to execute differently between `v1.8.6` and `v1.8.7`; this was originally discovered when testing `v1.8.9` release against testnet (majority on `v1.8.5`). For completeness, `[v1.8.5, v1.8.6]` behave correctly whereas `[v1.8.7, v1.8.8, v1.8.9]` all show the issue.

```
[INFO  solana_ledger_tool] solana-ledger-tool 1.8.6
[INFO  solana_runtime::bank] bank frozen: 107806633
    hash: 8qoFHi2xqbbQrEM4LeqXtmgLV3EetkTbpjZJcNs9AABv
    accounts_delta: 2woWGnggb3HePM3dGyxAnBrLfmGetX1t2RSwhJQXzke9
    signature_count: 3563
    last_blockhash: GFwPvnkAzFtziMHJuANxYTQBBxeYpusjJ9TpE64vpzVq
    capitalization: 1428634255458376098
[INFO  solana_runtime::bank] accounts hash slot: 107806633
    stats: BankHashStats {
        num_updated_accounts: 6329,
        num_removed_accounts: 1,
        num_lamports_stored: 558183844229069922,
        total_data_len: 12041969,
        num_executable_accounts: 1 }

[INFO  solana_ledger_tool] solana-ledger-tool 1.8.7
[INFO  solana_runtime::bank] bank frozen: 107806633
    hash: 5un5a6az2Z7HADcEK1cPQyBwag4PmTfYC2QqJVrcGbBa
    accounts_delta: G8iLX7YrctV7pFo8ZcmRVxhfvvyMtJSqxQCAAGmQXidJ
    signature_count: 3563
    last_blockhash: GFwPvnkAzFtziMHJuANxYTQBBxeYpusjJ9TpE64vpzVq
    capitalization: 1428634255458376098
[INFO  solana_runtime::bank] accounts hash slot: 107806633
    stats: BankHashStats {
        num_updated_accounts: 6327,
        num_removed_accounts: 0,
        num_lamports_stored: 558183833118473922,
        total_data_len: 10446011,
        num_executable_accounts: 0 }
```

Here is the information for that transaction (from slot `107806633`):

```
  Entry 40 - num_hashes: 4161, hashes: 6KHbZnr3L53GbqYiUGQTuh9iDcD7q25WGoD7Y3t4dgpX, transactions: 1
    Transaction 0
      Recent Blockhash: 9b3PsiYYsNAh2vMTWZ69892LLTDkrzheNcseKhGXxkr2
      Signature 0: 5idXTZb4poc4Wnyv3sh7pwrGc5n38T8MdwvWWasHn1VVyDdCeo7f11rK5TXMAj6fp3Ko6VEdCmkGiGWmbRcoB3wz
      Account 0: srw- HcHoaisCkRBoHn1kdbctRUFQTjbN47GYaRwu8VL8wLFW (fee payer)
      Account 1: -rw- 8W7g9xo4Ea67hheZEteD1VFVi1SoBy8yv95ZXoj6DkRg
      Account 2: -rw- 9J7VHshPZ9G8hhfcmrQfVpNQvZMLpzMjGJEqX3qyqPDB
      Account 3: -rw- EEKEhJWcwcJ1GtE9B27naztmx2393WtFyV8AXsWAjpfM
      Account 4: -r-- SysvarRent111111111111111111111111111111111
      Account 5: -r-- SysvarC1ock11111111111111111111111111111111
      Account 6: -r-x BPFLoaderUpgradeab1e11111111111111111111111
      Instruction 0
        Program:   BPFLoaderUpgradeab1e11111111111111111111111 (6)
        Account 0: 8W7g9xo4Ea67hheZEteD1VFVi1SoBy8yv95ZXoj6DkRg (1)
        Account 1: 9J7VHshPZ9G8hhfcmrQfVpNQvZMLpzMjGJEqX3qyqPDB (2)
        Account 2: EEKEhJWcwcJ1GtE9B27naztmx2393WtFyV8AXsWAjpfM (3)
        Account 3: HcHoaisCkRBoHn1kdbctRUFQTjbN47GYaRwu8VL8wLFW (0)
        Account 4: SysvarRent111111111111111111111111111111111 (4)
        Account 5: SysvarC1ock11111111111111111111111111111111 (5)
        Account 6: HcHoaisCkRBoHn1kdbctRUFQTjbN47GYaRwu8VL8wLFW (0)
        Data: [3, 0, 0, 0]
      Status: Unavailable
```

And here is the debug output for that transaction:
```
[DEBUG solana_runtime::message_processor] ELF error: Found .bss section in ELF, read-write data not supported
[DEBUG solana_runtime::bank] check: 4us load: 1390us execute: 601us txs_len=1
[DEBUG solana_runtime::bank] tx error: Err(InstructionError(0, InvalidAccountData))
    Transaction {
        signatures:[5idXTZb4poc4Wnyv3sh7pwrGc5n38T8MdwvWWasHn1VVyDdCeo7f11rK5TXMAj6fp3Ko6VEdCmkGiGWmbRcoB3wz],
        message: Message {
        header: MessageHeader {
            num_required_signatures: 1,
            num_readonly_signed_accounts: 0,
            num_readonly_unsigned_accounts: 3 },
            account_keys: [HcHoaisCkRBoHn1kdbctRUFQTjbN47GYaRwu8VL8wLFW, 8W7g9xo4Ea67hheZEteD1VFVi1SoBy8yv95ZXoj6DkRg, 9J7VHshPZ9G8hhfcmrQfVpNQvZMLpzMjGJEqX3qyqPDB, EEKEhJWcwcJ1GtE9B27naztmx2393WtFyV8AXsWAjpfM, SysvarRent111111111111111111111111111111111, SysvarC1ock11111111111111111111111111111111, BPFLoaderUpgradeab1e11111111111111111111111],
            recent_blockhash: 9b3PsiYYsNAh2vMTWZ69892LLTDkrzheNcseKhGXxkr2,
            instructions: [CompiledInstruction { program_id_index: 6, accounts: [1, 2, 3, 0, 4, 5, 0], data: [3, 0, 0, 0] }] } }
```

#### Steps To Reproduce 

1) Grab and expand this [ledger](https://drive.google.com/file/d/18c1e-5dU6Crx6i0zjSC4Aftw94k_jlTA/view?usp=sharing)
2) Run `solana-ledger-tool verify --ledger ~/21850_ledger/`
3) Note the `bank frozen: 107806633 hash: ...` log output

**Comment by Lichtso** [comment]
@jackcmay This is the one that https://github.com/solana-labs/rbpf/pull/240 should fix right?

**Comment by jackcmay** [comment]
Correct

**Comment by Lichtso** [comment]
Should be fixed by #21880, @steviez or @jackcmay can you confirm?

**Comment by steviez** [comment]
@Lichtso @jackcmay - With the tip of master, I confirmed that the wrong bank hash was computed without 509bcd2e742e158ae14bface4aeefbb37130ec8f (v0.2.19 upgrade commit) and that the right bank hash was calculated with that commit. So, yes, this issue is resolved by #21880
